streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -571,68 +630,155 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
                 partitionsForTask,
                 partitionsByHostState,
                 futureConsumers,
+                allOwnedPartitions,
                 minReceivedMetadataVersion
             );
         } else {
             assignment = computeNewAssignment(
                 clientMetadataMap,
                 partitionsForTask,
                 partitionsByHostState,
+                allOwnedPartitions,
                 minReceivedMetadataVersion
             );
         }
 
         return new GroupAssignment(assignment);
     }
 
-    private static Map<String, Assignment> computeNewAssignment(final Map<UUID, ClientMetadata> clientsMetadata,
-                                                                final Map<TaskId, Set<TopicPartition>> partitionsForTask,
-                                                                final Map<HostInfo, Set<TopicPartition>> partitionsByHostState,
-                                                                final int minUserMetadataVersion) {
+    private Map<String, Assignment> computeNewAssignment(final Map<UUID, ClientMetadata> clientsMetadata,
+                                                         final Map<TaskId, Set<TopicPartition>> partitionsForTask,
+                                                         final Map<HostInfo, Set<TopicPartition>> partitionsByHostState,
+                                                         final Set<TopicPartition> allOwnedPartitions,
+                                                         final int minUserMetadataVersion) {
         final Map<String, Assignment> assignment = new HashMap<>();
+        rebalanceRequired = false;
```

Comment:
nit: I think this variable can be reduced to a local one since it's only needed inside `computeNewAssignment`: before loop over the client initialize it to false. And then let `giveTasksBackToConsumers` to return empty map if it needs to beak fail fast (maybe rename to `tryGiveTasksBackToConsumsers`). And the caller becomes:

```
if (rebalanceRequired || state.ownedPartitions().isEmpty())
    assignment = interleaveTasksByGroupId
else if ((assignment = giveTasksBackToConsumers).equals(empty))
    assignment = interleaveTasksByGroupId
    rebalanceRequired = true
```

WDYT?

Reply:
Sounds good